baybe/searchspace/discrete.py

```diff
@@ -230,6 +231,122 @@ def discrete_parameter_factory(
+            which the parameters are passed to this method, as the configuration space
+            is built up incrementally from the parameter sequence.
+        """
+        # Separate parameter into numerical and other
+        numerical_parameters = [
+            p for p in parameters if isinstance(p, NumericalDiscreteParameter)
+        ]
+        other_parameters = [
+            p for p in parameters if not isinstance(p, NumericalDiscreteParameter)
+        ]
+
+        # Validate non-negativity
+        min_values = [min(p.values) for p in numerical_parameters]
+        if not (min(min_values) >= 0.0):
+            raise ValueError(
+                f"All parameters passed to '{cls.from_simplex.__name__}' "
+                f"must have non-negative values only."
+            )
+
+        def drop_invalid(df: pd.DataFrame, total: float, boundary_only: bool) -> None:
+            """Drop rows that violate a specified simplex constraint.
+
+            Args:
+                df: The dataframe whose rows should satisfy the simplex constraint.
+                total: The maximum row sum defining the simplex size.
+                boundary_only: Flag to control if the points represented by the rows
+                    may lie inside the simplex or on its boundary only.
+            """
+            row_sums = df.sum(axis=1)
+            if boundary_only:
+                rows_to_drop = row_sums[
+                    (row_sums < total - tolerance) | (row_sums > total + tolerance)
+                ].index
+            else:
+                rows_to_drop = row_sums[row_sums > total + tolerance].index
+            df.drop(rows_to_drop, inplace=True)
+
+        # Get the minimum sum contributions to come in the upcoming joins (the first
+        # item is the minimum possible sum of all parameters starting from the
+        # second parameter, the second item is the minimum possible sum starting from
+        # the third parameter, and so on ...)
+        min_upcoming = np.cumsum(min_values[:0:-1])[::-1]
+
+        # Incrementally build up the space, dropping invalid configuration along the
+        # way. More specifically: after having cross-joined a new parameter, there must
+        # be enough "room" left for the remaining parameters to fit. Hence,
+        # configurations of the current parameter subset that exceed the desired
+        # total value minus the minimum contribution to come from the yet to be added
+        # parameters can be already discarded.
+        for i, (param, min_to_go) in enumerate(
+            zip_longest(numerical_parameters, min_upcoming, fillvalue=0)
+        ):
+            if i == 0:
+                exp_rep = pd.DataFrame({param.name: param.values})
+            else:
+                exp_rep = pd.merge(
+                    exp_rep, pd.DataFrame({param.name: param.values}), how="cross"
+                )
+            drop_invalid(exp_rep, total - min_to_go, boundary_only=False)
```

Comment:
The way that you use `drop_invalid` here does not fit its docstring, since the second item is not `total` anymore.

Reply:
Can you elaborate what exactly you mean? Note btw that the "total" defined in this function locally does NOT need to coincide with the "total" defined in the outer scope. Therefore, it still DOES implement a simplex constraint, though a different simplex than the outside one. Does this resolve the issue? Anyway, could you just have a look at the latest version and let me know if you still think it's an issue?

Reply:
Ok, with the interpretation of this being technically a different simplex this makes sense.